arch/x86_64/mcount-insn.c

```diff
@@ -58,6 +58,245 @@ void print_instrument_fail_msg(int reason)
 	}
 }
 
+/*
+ *  handle PIC code.
+ *  for currently, this function targeted specific type of instruction.
+ *
+ *  this function manipulate the instruction like below,
+ *  => lea rcx, qword ptr [rip + 0x8f3f85]
+ *  to this.
+ *  => mov rcx, [address where instruction located actually]
+ *  => lea rcx, qword ptr [rcx + 0x8f3f85]
+ */
+static int handle_pic(cs_insn *insn, uint8_t insns[])
+{
+	cs_x86 *x86;
+	x86 = &insn->detail->x86;
+
+	/*
+	 * array for mov instruction.
+	 * ex) mov rbx, 0x555556d35690
+	 */
+	uint8_t mov_insns[] = {
+		0x48, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
+	};
+	int rex_offset = 0;
+	int reg_offset = 1;
+	int imm64_offset = 2;
+
+	/*
+	 * array for lea instruction.
+	 * ex) lea rbx, [rbx+0x8f3f85]
+	 */
+	uint8_t lea_insns[] = {
+		0x48, 0x8d, 0x00, 0x00, 0x00, 0x00, 0x00
+	};
+	int lea_rex_offset = 0;
+	int lea_insn_offset = 1;
+	int lea_opnd_offset = 2;
+	int lea_disp_offset = 3;
+
+	// REX.48 operands
+	uint8_t mov_opnd_48[] = {
+		//rax,	rbx,	rcx,	rdx,	rdi,	rsi,	rbp,	rsp
+		0xb8,	0xbb,	0xb9,	0xba,	0xbf,	0xbe,	0xbd,	0xbc
+	};
+
+	// REX.49 operands
+	uint8_t mov_opnd_49[] = {
+		//r8,	r9,	r10,	r11,	r12,	r13,	r14,	r15
+		0xb8,	0xb9,	0xba,	0xbb,	0xbc,	0xbd,	0xbe,	0xbf
+	};
+
+	// REX.48 operands,
+	uint8_t lea_opnd_48[] = {
+		//rax,	rbx,	rcx,	rdx,	rdi,	rsi,	rbp,	rsp
+		0x80,	0x9b,	0x89,	0x92,	0xbf,	0xb6,	0xad,	0xa4
+	};
+
+	// REX.4C operands,
+	uint8_t lea_opnd_4C[] = {
+		//r8,	r9,	r10,	r11,	r12,	r13,	r14,	r15
+		0x80,	0x89,	0x92,	0x9b,	0xa4,	0xad,	0xb6,	0xbf
+	};
+
+	// for currently, support LEA instruction only.
+	if (strncmp(insn->mnemonic, "lea", 3) != 0)
+		goto out;
+
+	// according to intel manual, lea instruction takes 2 operand only.
+	cs_x86_op *opnd1 = &(x86->operands[0]);
+	cs_x86_op *opnd2 = &(x86->operands[1]);
+
+	// get the register from opnd1 to load rip address
+	if (X86_REG_RAX <= opnd1->reg && opnd1->reg <= X86_REG_RSP) {
+		mov_insns[rex_offset] = 0x48;
+		lea_insns[lea_rex_offset] = 0x48;
+	}
+
+	else if (X86_REG_R8 <= opnd1->reg && opnd1->reg <= X86_REG_R15) {
+		mov_insns[rex_offset] = 0x49;
+		mov_insns[lea_rex_offset] = 0x4C;
+	}
+	else
+		goto out;
+
+	switch(opnd1->reg) {
+		// REX.48
+		case X86_REG_RAX:
+			mov_insns[reg_offset] = mov_opnd_48[0];
+			break;
+		case X86_REG_RBX:
+			mov_insns[reg_offset] = mov_opnd_48[1];
+			break;
+		case X86_REG_RCX:
+			mov_insns[reg_offset] = mov_opnd_48[2];
+			break;
+		case X86_REG_RDX:
+			mov_insns[reg_offset] = mov_opnd_48[3];
+			break;
+		case X86_REG_RDI:
+			mov_insns[reg_offset] = mov_opnd_48[4];
+			break;
+		case X86_REG_RSI:
+			mov_insns[reg_offset] = mov_opnd_48[5];
+			break;
+		case X86_REG_RBP:
+			mov_insns[reg_offset] = mov_opnd_48[6];
+			break;
+		case X86_REG_RSP:
+			mov_insns[reg_offset] = mov_opnd_48[7];
+			break;
```

Comment:
We don't indent `case` and it seems you can remove duplication here using reg -> index table.

Reply:
moved indent.